src/chat.rs

```diff
@@ -532,19 +532,59 @@ impl ChatId {
     }
 
     // Unarchives a chat that is archived and not muted.
-    // Needed when a message is added to a chat so that the chat gets a normal visibility again.
+    // Needed after a message is added to a chat so that the chat gets a normal visibility again.
     // Sending an appropriate event is up to the caller.
-    pub async fn unarchive_if_not_muted(self, context: &Context) -> Result<()> {
+    // Also emits DC_EVENT_MSGS_CHANGED for DC_CHAT_ID_ARCHIVED_LINK when the number of archived
+    // chats with unread messages increases (which is possible if the chat is muted).
+    pub async fn unarchive_if_not_muted(
+        self,
+        context: &Context,
+        msg_state: MessageState,
+    ) -> Result<()> {
+        if msg_state != MessageState::InFresh && msg_state != MessageState::InNoticed {
+            context.sql.execute(
+                "UPDATE chats SET archived=0 WHERE id=? AND NOT(muted_until=-1 OR muted_until>?)",
+                paramsv![self, time()],
+            ).await?;
+            return Ok(());
+        }
+        let chat = Chat::load_from_db(context, self).await?;
+        if chat.visibility != ChatVisibility::Archived {
+            return Ok(());
+        }
+        if chat.is_muted() {
+            let nr_unread = context
+                .sql
+                .count(
+                    "SELECT COUNT(*)
+                FROM msgs
+                WHERE (state=? OR state=?)
+                AND hidden=0
+                AND chat_id=?;",
+                    paramsv![MessageState::InFresh, MessageState::InNoticed, self],
```

Comment:
i think, it is wrong to take `MessageState::InNoticed` into account here.

only "fresh" messages are returned by `get_fresh_msg_cnt()`, "noticed" messages are no longer counted as "fresh" and are not shown in the "unread" badge.

see [here](https://c.delta.chat/classdc__msg__t.html#a83fbf6e74d09a0b598ccefe9b48bd68c) for some more details about fresh/noticed/seen.

Reply:
Thanks for pointing this out! It was a question for me. Looks like i misread the documentation